docs/Userguide_portability.rst

```diff
@@ -0,0 +1,67 @@
+Portability concerns and solutions
+==================================
+
+When working on a software project, it is important to be aware of all the
+software and libraries the project rely on and to list them explicitly and
+*under version control* in such a way that they can easily be installed and made
+available on different systems. The upsides are significant:
+
+* Easily install and run on the cluster
+* Ease of collaboration
+* Better reproducibility
+
+To achieve this, try to always keep in mind the following aspects:
+
+* **Versions:** For each dependency, make sure you have some record of the
+  specific version you are using during development. That way, in the future, you
+  will be able to reproduce the original environment which you know to be
+  compatible. Indeed, the more time passes, the more likely it is that newer
+  versions of some dependency have breaking changes. The ``pip freeze`` command can create
+  such a record for Python dependencies.
+* **Isolation:** Ideally, each of your software projects should be isolated from
+  the others. What this means is that updating the environment for project A
+  should *not* update the environment for project B. That way, you can freely
+  install and upgrade software and libraries for the former without worrying about
+  breaking the latter (which you might not notice until weeks later, the next time
+  you work on project B!) Isolation can be achieved using :ref:`Python Virtual
+  environments` and :ref:`containers`.
+
+.. Creating a list of your software's dependencies
+.. -----------------------------------------------
+.. TODO
+
+
+Managing your environments
+--------------------------
+
+.. include:: Userguide_python.rst
+
+
+Using Modules
+-------------
+
+A lot of software, such as Python and Conda, is already compiled and available on
+the cluster through the ``module`` command and its sub-commands. In particular,
+if you wish to use ``Python 3.7`` you can simply do:
+
+.. prompt:: bash $
+
+    module load python/3.7
+
+.. include:: Userguide_portability_modules.rst
+
+
+On using containers
+-------------------
+
+Another option for creating portable code is :ref:`Using containers`.
```

Comment:
Reported by @tesfaldet:
> ![Screen Shot 2021-08-07 at 5 27 12 PM](https://user-images.githubusercontent.com/5014902/128614256-eb904c94-92f9-43e6-9345-19b3ed5422cb.png)
> "Using" -> "using"

@abergeron the ref:`...` is broken in here: https://mila-docs-staging--46.org.readthedocs.build/en/46/Userguide.html#on-using-containers. Same goes here: https://mila-docs-staging--46.org.readthedocs.build/en/46/Theory_cluster.html#containers but what's weird is that the :ref: in the paragraph just above works: https://mila-docs-staging--46.org.readthedocs.build/en/46/Theory_cluster.html#cluster-software-modules. Would you know how to fix it?

Reply:
The ref is done wrong, it should reference a label, not a section title in the source.  So there should be a label (`.. _label_text:`) before the `Using Containers` section and the reference should read
```
:ref:`_label_text`
```

There is another option for file-local references that would read
```
`Using Containers`_
```
but that doesn't work for targets in a different file (source file, not output file).

Reply:
Also apparently "included" files count as the same source file, so everything within the "Userguide" section will work, but other will not.